core/src/main/java/com/linecorp/armeria/common/stream/TimeoutSubscriber.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import static java.util.Objects.requireNonNull;
+
+import java.time.Duration;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import io.netty.util.concurrent.EventExecutor;
+import io.netty.util.concurrent.ScheduledFuture;
+
+public class TimeoutSubscriber<T> implements Subscriber<T> {
+    private static final String TIMEOUT_MESSAGE = "Stream timed out after %d ms (timeout mode: %s)";
+    private final Subscriber<? super T> delegate;
+    private final EventExecutor executor;
+    private final StreamTimeoutMode timeoutMode;
+    private final Duration timeoutDuration;
+    private ScheduledFuture<?> timeoutFuture;
+    private Runnable timeoutTask;
+    private Subscription subscription;
+    private long timeoutNanos;
+    private long lastOnNextTimeNanos;
+
+    public TimeoutSubscriber(Subscriber<? super T> delegate, EventExecutor executor, Duration timeoutDuration, StreamTimeoutMode timeoutMode) {
+        this.delegate = requireNonNull(delegate, "delegate");
+        this.executor = requireNonNull(executor, "executor");
+        this.timeoutDuration = requireNonNull(timeoutDuration, "timeoutDuration");
+        timeoutNanos = timeoutDuration.toNanos();
+        this.timeoutMode = requireNonNull(timeoutMode, "timeoutMode");
+        timeoutTask = createTimeoutTask();
+    }
+
+    private Runnable createTimeoutTask() {
+        return () -> {
+            if(timeoutMode == StreamTimeoutMode.UNTIL_NEXT) {
+                long currentTimeNanos = System.nanoTime();
+                long elapsedNanos = currentTimeNanos - lastOnNextTimeNanos;
+
+                if(elapsedNanos <= timeoutNanos) {
+                    long delayNanos = timeoutNanos - (currentTimeNanos - lastOnNextTimeNanos);
+                    timeoutFuture = createTimeoutSchedule(delayNanos, TimeUnit.NANOSECONDS);
+                    return;
+                }
+            }
+            subscription.cancel();
+            delegate.onError(new TimeoutException(String.format(TIMEOUT_MESSAGE, timeoutDuration.toMillis(), timeoutMode)));
```

Comment:
There is no guarantee that `subscription.cancel` will synchronize with upstream to not send additional signals.

e.g.`delegate.onError` could be called twice, once from L65 and once from L108, or `delegate.onError` (L65) and `delegate.onComplete` (L114) could be called.

> If a Subscription is cancelled its Subscriber MUST eventually stop being signaled ... The reason for eventually is because signals can have propagation delay due to being asynchronous.

ref: https://github.com/reactive-streams/reactive-streams-jvm?tab=readme-ov-file#1.8
ref: https://github.com/reactive-streams/reactive-streams-jvm?tab=readme-ov-file#1.7